core/o11y/src/lib.rs

```diff
@@ -82,21 +94,55 @@ pub fn default_subscriber(
     let stderr = std::io::stderr();
     let lined_stderr = std::io::LineWriter::new(stderr);
     let (writer, writer_guard) = tracing_appender::non_blocking(lined_stderr);
-    let subscriber = tracing_subscriber::FmtSubscriber::builder()
+
+    let subscriber_builder = tracing_subscriber::FmtSubscriber::builder()
         .with_span_events(
             tracing_subscriber::fmt::format::FmtSpan::ENTER
                 | tracing_subscriber::fmt::format::FmtSpan::CLOSE,
         )
-        .with_env_filter(log_filter)
         .with_writer(writer)
-        .finish();
+        .with_env_filter(log_filter)
+        .with_filter_reloading();
+    let reload_handle = subscriber_builder.reload_handle();
+    ENV_FILTER_RELOAD_HANDLE.set(reload_handle).unwrap();
+
+    let subscriber = subscriber_builder.finish();
     DefaultSubcriberGuard {
         subscriber: Some(subscriber),
         local_subscriber_guard: None,
         writer_guard,
     }
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum ReloadError {
+    #[error("reload error")]
+    ReloadErrorType { err: Error },
+    #[error("env_filter directive parse error")]
+    ParseErrorType { err: ParseError },
+    #[error("env_filter reload handle is not available")]
+    NoReloadHandlerType,
+}
+
+impl From<ParseError> for ReloadError {
+    fn from(err: ParseError) -> Self {
+        Self::ParseErrorType { err }
+    }
+}
+
+impl From<tracing_subscriber::reload::Error> for ReloadError {
+    fn from(err: Error) -> Self {
+        Self::ReloadErrorType { err }
+    }
+}
+
+pub fn reload_env_filter(rust_log: &str) -> Result<(), ReloadError> {
+    ENV_FILTER_RELOAD_HANDLE.get().map_or(Err(ReloadError::NoReloadHandlerType), |reload_handle| {
+        reload_handle.reload(EnvFilter::from_str(rust_log)?)?;
```

Comment:
I think it would be better to adjust/refactor `EnvFilterBuilder` to allow for the use-cases needed by this.

In particular one problem that `EnvFilterBuilder` solves is setting up logging in such a way that some 3rd party libraries such as `h2` are filtered out by default – and those tend to be _very_ chatty.

It might very well be the case that enabling `h2` logging from this endpoint is actually desirable, but I'm also worried slightly that people will be quite frustrated about specifying `debug`, finding out that logs are unusable, and having to retry after specifying something like `debug,h2=warn,trust_dns_resolver=warn...`.

Reply:
Updated to provide the same behavior as the command-line provides.